benchmark/cajita/Cajita_HaloPerformance.cpp

```diff
@@ -0,0 +1,274 @@
+
+    // Get comm rank;
+    int comm_rank;
+    MPI_Comm_rank( MPI_COMM_WORLD, &comm_rank );
+
+    // Get comm size;
+    int comm_size;
+    MPI_Comm_size( MPI_COMM_WORLD, &comm_size );
+
+    // Get Cartesian comm
+    std::array<int, 3> ranks_per_dim;
+    for ( std::size_t d = 0; d < 3; ++d )
+        ranks_per_dim[d] = 0;
+    MPI_Dims_create( comm_size, 3, ranks_per_dim.data() );
+
+    int global_cells_per_dim_size = problem_global_cells_per_dim.size();
+    for ( int p = 0; p < global_cells_per_dim_size; ++p )
+    {
+        // Open the output file on rank 0.
+        std::fstream file;
+        if ( 0 == comm_rank )
+            file.open( filename + "_" + std::to_string( comm_size ) + "_" +
+                           std::to_string( problem_global_cells_per_dim[p] ),
+                       std::fstream::out );
+
+        // Output problem details.
+        if ( 0 == comm_rank )
+        {
+            file << "\n";
+            file << "Cajita Halo Performance Benchmark"
+                 << "\n";
+            file << "----------------------------------------------"
+                 << "\n";
+            file << "MPI Ranks: " << comm_size << "\n";
+            file << "MPI Cartesian Dim Ranks: (" << ranks_per_dim[0] << ", "
+                 << ranks_per_dim[1] << ", " << ranks_per_dim[2] << ")\n";
+            file << "----------------------------------------------"
+                 << "\n";
+            file << "\n";
+        }
+
+        // Do everything on the CPU.
+#ifdef KOKKOS_ENABLE_SERIAL
+        using SerialDevice = Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>;
+        performanceTest<SerialDevice>( file, problem_global_cells_per_dim,
+                                       "serial_", problem_halo_width,
+                                       MPI_COMM_WORLD );
+#endif
+#ifdef KOKKOS_ENABLE_OPENMP
+        using OpenMPDevice = Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>;
+        performanceTest<OpenMPDevice>( file, problem_global_cells_per_dim,
+                                       "openmp_", problem_halo_width,
+                                       MPI_COMM_WORLD );
+#endif
+
+#ifdef KOKKOS_ENABLE_CUDA
+        using CudaDevice = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>;
+        // Do everything on the GPU with regular GPU memory.
+        performanceTest<CudaDevice>( file, problem_global_cells_per_dim,
+                                     "cuda_", problem_halo_width,
+                                     MPI_COMM_WORLD );
+#endif
+
+#ifdef KOKKOS_ENABLE_HIP
+        using HipDevice = Kokkos::Device<Kokkos::Experimental::HIP,
+                                         Kokkos::Experimental::HIPSpace>;
+        performanceTest<HipDevice>( file, problem_global_cells_per_dim, "hip_",
+                                    problem_halo_width, MPI_COMM_WORLD );
+#endif
+
+        // Close the output file on rank 0.
+        if ( 0 == comm_rank )
+            file.close();
```

Comment:
```suggestion
```
You can keep it if you really want to but it is unnecessary, since the file-based stream goes out of scope and the destructor will close the file.

Reply:
deleted